lib/src/rules/avoid_unused_constructor_parameters.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Avoid defining unused paramters in constructors.';
+
+const _details = r'''
+
+**AVOID** defining unused parameters in constructors.
+
+**BAD:**
+```
+class BadOne {
+  BadOne(int unusedParameter, [String unusedPositional]);
+}
+
+class BadTwo {
+  int c;
+
+  BadTwo(int a, int b, int x) {
+    c = a + b;
+  }
+}
+```
+
+''';
+
+class AvoidUnusedConstructorParameters extends LintRule {
+  _Visitor _visitor;
+  AvoidUnusedConstructorParameters()
+      : super(
+            name: 'avoid_unused_constructor_parameters',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitConstructorDeclaration(ConstructorDeclaration node) {
+    final _constructorVisitor = new _ConstructorVisitor(rule, node);
+    node?.body?.visitChildren(_constructorVisitor);
+    node?.initializers?.forEach((i) => i.visitChildren(_constructorVisitor));
+
+    final unusedParameters = _constructorVisitor.unusedParameters;
+    if (unusedParameters.isNotEmpty) rule.reportLint(unusedParameters.first);
```

Comment:
It would be better to report every unused parameter, rather than only the first.

Reply:
The test suite expects one error from lines tagged with `// LINT`. Should I just report the entire line? Or is there a way I can get the test suite to expect multiple lint reports?

Reply:
In cases like that, however artificial it looks, I'd just put each offending statement on its own line.  For example:

```dart

class A {
  A(int x,  // LINT
    int y); // LINT
}
```